src/libcprover-cpp/verification_result.cpp

```diff
@@ -0,0 +1,150 @@
+{
+  propertiest _properties;
+  resultt _verifier_result;
+
+public:
+  explicit verification_result_implt() = default;
+  verification_result_implt(const verification_result_implt &other) = default;
+  ~verification_result_implt() = default;
+
+  void set_properties(propertiest &properties);
+  void set_result(resultt &verification_result);
+
+  resultt get_result();
+  const propertiest &get_properties();
+};
+
+// Impl
+
+void verification_resultt::verification_result_implt::set_properties(
+  propertiest &properties)
+{
+  _properties = properties;
+}
+
+void verification_resultt::verification_result_implt::set_result(
+  resultt &result)
+{
+  _verifier_result = result;
+}
+
+resultt verification_resultt::verification_result_implt::get_result()
+{
+  return _verifier_result;
+}
+
+const propertiest &
+verification_resultt::verification_result_implt::get_properties()
+{
+  return _properties;
+}
+
+// Verification_result
+
+verification_resultt::verification_resultt()
+  : _impl(util_make_unique<verification_resultt::verification_result_implt>())
+{
+}
+
+verification_resultt::~verification_resultt()
+{
+}
+
+verification_resultt::verification_resultt(const verification_resultt &other)
+  : _impl(util_make_unique<verification_result_implt>(*other._impl))
+{
+}
+
+verification_resultt &
+verification_resultt::operator=(verification_resultt &&) = default;
+verification_resultt &
+verification_resultt::operator=(const verification_resultt &&other)
+{
+  *_impl = *other._impl;
+  return *this;
+}
+
+void verification_resultt::set_properties(propertiest &properties)
+{
+  _impl->set_properties(properties);
+}
+
+void verification_resultt::set_result(resultt &result)
+{
+  _impl->set_result(result);
+}
+
+verifier_resultt verification_resultt::final_result() const
+{
+  switch(_impl->get_result())
+  {
+  case resultt::ERROR:
+    return verifier_resultt::ERROR;
+  case resultt::FAIL:
+    return verifier_resultt::FAIL;
+  case resultt::PASS:
+    return verifier_resultt::PASS;
+  case resultt::UNKNOWN:
+    return verifier_resultt::UNKNOWN;
+  }
+  // Required to silence compiler warnings that are promoted as errors!
+  UNREACHABLE;
+}
+
+std::vector<std::string> verification_resultt::get_property_ids() const
+{
+  std::vector<std::string> result;
+  for(const auto &props : _impl->get_properties())
+  {
+    result.push_back(as_string(props.first));
+  }
+  return result;
+}
+
+std::string verification_resultt::get_property_description(
+  const std::string &property_id) const
+{
+  auto dstringt_property = dstringt(property_id);
+  const auto description =
+    _impl->get_properties().at(dstringt_property).description;
+  return description;
+}
+
+prop_statust
+verification_resultt::get_property_status(const std::string &property_id) const
+{
+  auto dstringt_property = dstringt(property_id);
+  switch(_impl->get_properties().at(dstringt_property).status)
+  {
+  case property_statust::ERROR:
```

Comment:
:bulb: I suggest making the error case the default case. This will avoid compile errors about reaching the end of non-void functions.

Reply:
I prefer to keep the `UNREACHABLE` - this is to communicate intent about this being an exhaustive list.

It also guards against a new enum member being added but not being matched against properly, as it will fail fast and loud.

Reply:
As I'd previously pointed out in one of my other comments on this PR, the standard `INVARIANT` / `UNREACHABLE` macros don't work in a useful fashion in this context. Therefore an alternative means of reporting the error is required. I suggest using a comment if you feel the need to explain the the default case should be unreachable and is therefore an error.